### PR TITLE
Add configurable storage class name for image resource

### PR DESCRIPTION
**Pull request by tylerphelan**
- If set, will be used as the storageClassName in the pvc

### CODE REVIEW

**Review by jjbustamante — COMMENTED** [review]
It looks good to me!

**Comment by tomkennedy513** [comment]
I'm wondering if we actually need to validate that the storage class exists since that will require additional rbac for our controller. If we create a pod with an invalid storage class can't we just let pod fail?

File: pkg/reconciler/image/image.go [hunk 1]
```diff
@@ -145,6 +151,15 @@ func (c *Reconciler) reconcileImage(ctx context.Context, image *buildapi.Image)
 		return image, nil
 	}
 
+	if image.Spec.NeedVolumeCache() && image.Spec.Cache.Volume.StorageClassName != "" {
+		_, err := c.StorageClassLister.Get(image.Spec.Cache.Volume.StorageClassName)
+		if err != nil {
```
Comment (matthewmcnew) [inline pkg/reconciler/image/image.go:156]: What is the reason to add this validation?

Comment (tylerphelan) [reply pkg/reconciler/image/image.go:156]: @matthewmcnew @tomkennedy513 

I think this is an error that would only show up in the kpack-controller logs and result in a pending Image which is an error we should bubble up more

Comment (tomkennedy513) [reply pkg/reconciler/image/image.go:156]: could we add it to the status of the image

Comment (tylerphelan) [reply pkg/reconciler/image/image.go:156]: > could we add it to the status of the image

That was my intention. I think it should be ready: false with a message

Comment (tomkennedy513) [reply pkg/reconciler/image/image.go:156]: ya I think that works. we should just make sure we document the required rbac change to the kpack controller

**Review by tomkennedy513 — APPROVED** [review]

File: pkg/reconciler/image/image.go (hunk 1 above)
Comment (matthewmcnew) [reply pkg/reconciler/image/image.go:156]: > I think this is an error that would only show up in the kpack-controller logs and result in a pending Image which is an error we should bubble up more

I am not sure I understand. Won't this be written to the status and prevent future reonciles?

Comment (tomkennedy513) [reply pkg/reconciler/image/image.go:156]: I don't think the pvc will enter a failed state if the storage class does not exist, it just stays pending

Comment (matthewmcnew) [reply pkg/reconciler/image/image.go:156]: After pondering this for a few hours, I don't think it makes sense for the image reconciler to validate what is and is not a valid storageClass or persistent volume configuration. It seems that should be the responsibility of the PersistentVolumeClaim resource.

Comment (tylerphelan) [reply pkg/reconciler/image/image.go]: removed!

**Review by chenbh — APPROVED** [review]

**Comment by matthewmcnew** [comment]
Does this need to be supported in v1alpha1 apis? https://github.com/pivotal/kpack/pull/934#issuecomment-1063374631

**Comment by tylerphelan** [comment]
@matthewmcnew 
>Does this need to be supported in v1alpha1 apis? https://github.com/pivotal/kpack/pull/934#issuecomment-1063374631

converting v1alpha1 -> v1alpha2:
`storageClassName` is always "", v1alpha2 clients will use ""

converting v1alpha2 -> v1alpha1:
v1alpha1 client (old version of `kp`) patches existing v1alpha2 image, causing `storageClassName` to be set to "" which will violate the immutability constraint causing blocking error. This is not great.

My 2c is that if an old v1alpha1 client runs into an immutability error it will not lose data. If the storageClassName is unset in v1alpha2 behavior will not be affected by v1alpha1 clients. Users should upgrade kp.

**Comment by matthewmcnew** [comment]
> v1alpha1 client (old version of kp) patches existing v1alpha2 image, causing storageClassName to be set to "" which will violate the immutability constraint causing blocking error. This is not great.

This seems rare but, not great. It essentially means we are not supporting v1alpha1.

**Comment by tylerphelan** [comment]
>It essentially means we are not supporting v1alpha1.

risk seems very high

>If our strategy for v1alpha1 headaches is to say not use v1alpha1 we should probably just begin removing v1alpha1.

If we don't do this will we need to migrate all new fields to prevent data loss? ex `DefaultProcess`, `ProjectDescriptorPath`, `CreationTime` ?

**Comment by matthewmcnew** [comment]
> If we don't do this will we need to migrate all new fields to prevent data loss? ex DefaultProcess, ProjectDescriptorPath, CreationTime

We probably should provide support for these. Hitting an expected immutability constraint seems slightly worse to me.

**Comment by tylerphelan** [comment]
>Hitting an expected immutability constraint seems slightly worse to me.

I think data loss is equal if not worse

**Comment by tylerphelan** [comment]
I think we should probably add migrations

**Comment by matthewmcnew** [comment]
> I think we should probably add migrations

Do you think we should do that as this PR or create an issue to support it for all susceptible fields?

**Comment by tylerphelan** [comment]
I've added the conversion for this one here and have added https://github.com/pivotal/kpack/issues/1036

**Review by tomkennedy513 — APPROVED** [review]

**Comment by tylerphelan** [comment]
will merge after 0.7.1 is released

**Review by matthewmcnew — APPROVED** [review]

**Pull request merged**